modules/csg/editor/csg_gizmos.cpp

```diff
@@ -38,6 +38,135 @@
 #include "editor/plugins/gizmos/gizmo_3d_helper.h"
 #include "editor/plugins/node_3d_editor_plugin.h"
 #include "scene/3d/camera_3d.h"
+#include "scene/3d/mesh_instance_3d.h"
+#include "scene/3d/physics/collision_shape_3d.h"
+#include "scene/3d/physics/static_body_3d.h"
+#include "scene/gui/dialogs.h"
+#include "scene/gui/menu_button.h"
+
+void CSGShapeEditor::_node_removed(Node *p_node) {
+	if (p_node == node) {
+		node = nullptr;
+		options->hide();
+	}
+}
+
+void CSGShapeEditor::edit(CSGShape3D *p_csg_shape) {
+	node = p_csg_shape;
+	if (node) {
+		options->show();
+	} else {
+		options->hide();
+	}
+}
+
+void CSGShapeEditor::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_THEME_CHANGED: {
+			options->set_icon(get_editor_theme_icon(SNAME("CSGCombiner3D")));
+		} break;
+	}
+}
+
+void CSGShapeEditor::_menu_option(int p_option) {
+	Array meshes = node->get_meshes();
+	if (meshes.is_empty()) {
+		err_dialog->set_text(TTR("CSG operation returned empty!"));
```

Comment:
Empty what?

Reply:
There is this CSG chain quirk where `CSGShape3D::get_meshes()` returns an empty `Array()` when the CSG root mesh is not valid. Since this menu option is only called when a valid CSG root node is selected the only way to have an invalid root mesh is a broken CSG operation (e.g. brush null) that had an empty error return because it has unref() the root mesh already. If just the mesh would be empty it would not have an invalid root mesh.

Reply:
I meant it says "returned empty", but empty what? Array? Mesh? It should be in the message.